cpp/ql/test/library-tests/members/this/test.cpp

```diff
@@ -0,0 +1,56 @@
+
+int global;
+
+class C {
+  int x;
+
+public:
+
+  void f1() {
+    // Implicit dereference of `this.`
+    x++;
+  }
+
+  void f2() {
+    // Explicit dereference of `this.`
+    this->x++;
+  }
+
+  int f3() const {
+    // We expect the type of `this` to be const-qualified.
+    return x;
+  }
+
+  int f4() volatile {
+    // We expect the type of `this` to be volatile-qualified.
+    return x;
+  }
+
+  int f5() const volatile {
+    // We expect the type of `this` to be qualified as both const and volatile.
+    return x;
+  }
+
+  void f6() {
+    // No use of `this`, but we still expect to be able to get its type.
+    global++;
+  }
+
+  float f7() const & {
+    // We expect the type of `this` to be const-qualified.
+    return x;
+  }
+
+  float f8() && {
+    // We expect the type of `this` to be unqualified.
+    return x;
+  }
+};
+
+// We want to test that D* is in the database even when there's no use of it,
+// not even through an implicit dereference of `this`.
+class D {
+  void f() {
+    global++;
+  }
+};
```

Comment:
Might be worth adding a couple of template classes, one that gets instantiated and one that doesn't, too.

Reply:
I've added that. For the instantiated one, we get results for both the instantiation and the template. For the uninstantiated one, we get no results.